src/fees/hooks.ts

```diff
@@ -43,9 +43,60 @@ export function usePaidFees(fees: Fee[]) {
   }
 }
 
-// Returns the maximum amount a user can send, taking into acount gas fees required for the transaction
-// also optionally fetches new fee estimations if the current ones are missing or out of date
 export function useMaxSendAmount(
+  tokenId: string,
+  feeType: FeeType.SEND | FeeType.SWAP,
+  shouldRefresh: boolean = true
+) {
+  const dispatch = useDispatch()
+  const balance = useTokenInfo(tokenId)?.balance ?? new BigNumber(0)
+  const feeEstimates = useSelector(feeEstimatesSelector)
+  const tokenInfo = useTokenInfo(tokenId)
+
+  // Optionally Keep Fees Up to Date
+  useEffect(() => {
+    if (!shouldRefresh || !tokenInfo?.address) return
+    const feeEstimate = feeEstimates[tokenInfo.address]?.[feeType]
+    if (
+      (feeType === FeeType.SWAP && balance.gt(0)) ||
+      !feeEstimate ||
+      feeEstimate.error ||
+      feeEstimate.lastUpdated < Date.now() - ONE_HOUR_IN_MILLIS
+    ) {
+      dispatch(estimateFee({ feeType, tokenAddress: tokenInfo.address }))
+    }
+  }, [tokenId, shouldRefresh])
+
+  const celoAddress = useSelector(celoAddressSelector)
+
+  // useFeeCurrency chooses which crypto will be used to pay gas fees. It looks at the valid fee currencies (cUSD, cEUR, CELO)
+  // in order of highest balance and selects the first one that has more than a minimum threshhold of balance
+  // if CELO is selected then it actually returns undefined
+  const feeTokenAddress = useFeeCurrency() ?? celoAddress
+
+  const usdFeeEstimate = tokenInfo?.address
+    ? feeEstimates[tokenInfo.address]?.[feeType]?.usdFee
+    : undefined
+  const feeEstimate =
+    useUsdToTokenAmount(new BigNumber(usdFeeEstimate ?? 0), tokenInfo?.address ?? undefined) ??
+    new BigNumber(0)
+
+  if (!balance) {
+    return new BigNumber(0)
+  }
+  // For example, if you are sending cUSD but you have more CELO this will be true
+  if (tokenInfo?.address !== feeTokenAddress) {
+    return balance
+  }
+  return balance.minus(feeEstimate)
+}
+
+// Returns the maximum amount a user can send, taking into account gas fees required for the transaction
+// also optionally fetches new fee estimations if the current ones are missing or out of date
+/**
+ * @deprecated use useMaxSendAmount instead
+ */
+export function useMaxSendAmountByAddress(
```

Comment:
It looks like this largely duplicates `useMaxSendAmount`. That's not ideal for the following reasons:
- makes our code harder to understand, because there is more code to read through in total to understand how the two hooks work
- makes our code less safe from bugs, because if a bug crops up, it now needs to be solved in 2 different places

In general, I suggest challenging yourself to never copy/paste code from another part of the codebase. When you do that, chances are you're missing an opportunity to create some helper function that would be useful in two places. And since you're not writing each line, you might miss some detail that will lead to bugs.

In this case, you could refactor useMaxSendAmountByAddress to just be a wrapper for your new implementation, shrinking the implementation down to just 2 lines
```
  const tokenInfo = useTokenInfoByAddress(tokenAddress)
  return useMaxSendAmount(tokenInfo?.tokenId, feeType, shouldRefresh)
```

However, this would require `useMaxSendAmount` to have an optional `tokenId`, which in turn would require `useTokenInfo` to make `tokenId` optional. I think this is going to be useful in the end anyway, since porting all the usages of `useTokenInfoByAddress` (24 right now) over to `useTokenInfo` without making the input optional sounds like unnecessary scope. I'll let @jophish comment if he had some intention to make inputs of hooks required more often, but it seems cumbersome to me to require that, given that you often want to use hooks by other hooks, which being governed by rules of hooks (no conditional evaluation in particular), make it hard to guarantee that inputs will be truthy in general

Reply:
alternatively, you could just update the 2 usages of `useMaxSendAmountByAddress` to use your new version, since both usages appear to already have easy access to a tokenId. However, I think you will run into the same issue of `tokenId` being required in both cases. More evidence that `useTokenInfo` needs `tokenId` to be optional IMO

Reply:
I created this visual on the general strategic choice. When there's this much duplication between the celo-specific implementation and the multichain one, I don't see option 2 as viable. https://www.figma.com/file/iiPCJG2cA18ysZrLAZI80d/Multichain-refactor-strategies?type=whiteboard&node-id=0%3A1&t=157ni2SZLUNGpni9-1

Reply:
Updating the two usages of `useMaxSendAmountByAddress` looks like it would require updating `FiatExchangeAmount.tsx` and `SwapScreen.tsx` which means using tokenId in the swap flow which impacts RET-872. It seems easier to simply create a new `useMaxSendAmount` function which takes tokenId can be used by swaps when ready and migrate the flows to that new function as we add multi-chain support @jh2oman.

Reply:
Fixed in https://github.com/valora-inc/wallet/pull/4242/commits/d6786bd16d6cf578a3a1de8c45e1d57432028718!